Add reference-based gene symbol conversion

jashapiro opened the pull request:
After a number of fits and starts with how best to do the gene identifier conversion, I ended up deciding the best way to handle things was to use a set of defined references where the conversion between Ensembl ids and gene symbols could be made uniform and simpler. 

This PR includes two main parts: 
- a notebook for downloading example datasets from 10x and compiling the reference gene table
- modifications to the conversion scripts to use those references

The table construction happens in a notebook in the `data-raw`, which also includes some exploration of the conversions. I could probably move the exploration out to a separate notebook and just have a simple script for constructing the table; let me know if you think that would be a better approach. I also store the table only for internal use; I suppose I could make it available outside the package as well, but I wasn't sure whether that would be worth the effort. Maybe it would?

The reference table is built using both a 10x 2020 reference that corresponds to Ensembl 98, and the newly release 2024 reference from Ensembl 110, as well as the default which is the ScPCA reference. I also include both the "plain" gene symbol conversions and the results from `make.unique()`

For the conversion scripts, I kept the ability to convert based on an SCE object, but moved the primary conversion for ensembl id lists to be based on the reference table. In the case of converting an SCE object, I left the internal row rata as the default source for conversion). There are options for using each of the different references as well as whether to make the results unique (which will always use the precomputed `make.unique` values). Let me know if you think there are other options that would be good to include, or any thoughts you might have on how we might change this for easier use. 

I expect we will want to add a bit more to the docs about the different references and when you might favor one or another, but I wanted to get the code in its current state up for discussion.

sjspielman commented on data-raw/build_gene_reference.Rmd:
````diff
@@ -0,0 +1,179 @@
+
+How many of the gene symbols in the 10x 2020 reference are in the ScPCA data, but perhaps with a different Ensembl ID?
+
+```{r}
+symbols_10x2020 <- scpca_missing$gene_symbol_10x2020[!is.na(scpca_missing$gene_symbol_10x2020)]
+sum(symbols_10x2020 %in% genes_scpca$gene_symbol)
+```
+
+What are they?
+
+```{r}
+# Match the genes that are in the 10x 2020 reference to the ScPCA table on gene symbol
+# Only for those where the gene id is missing in the ScPCA table
+genes_10x2020 |>
+  filter(
+    gene_symbol_10x2020 %in% genes_scpca$gene_symbol,
+    !gene_ids %in% genes_scpca$gene_ids
+  ) |>
+  left_join(genes_scpca, by = c("gene_symbol_10x2020" = "gene_symbol_scpca"))
````
I'm confused about what's going on here - when I run the code (and what I get is the same as what's in the html you added), it comes up that there are 0 such gene symbols. The code after "What are they?" seems to suggest this 0 might be expected? Clearly it seems you saw `LINC01505` go by at one point while working on this code though, so my tentative guess is... you updated to a different link ("This gene seems to have simply been removed in later revisions, so I think we can safely not worry about it.") and already decided not to worry about it?

jashapiro replied:
I forgot to rename a variable when I did a little refactor. Should be fixed now.

sjspielman commented on data-raw/build_gene_reference.Rmd:
```diff
@@ -0,0 +1,179 @@
+
+### BAC removal
+
+In Ensembl version 104, the version we are using for ScPCA, BAC-based gene IDs were removed and replaced with simply the Ensembl gene ID.
```
One does not simply replace with the Ensembl gene ID.

this is a recreational-for-fun review comment as I do not mind the use of "simply" 😉 😅

sjspielman commented on data-raw/build_gene_reference.Rmd:
````diff
@@ -0,0 +1,179 @@
+```
+
+
+## Look at some stats for the tables
````
At some point later" I got a little bit lost in keeping in my brain "are we checking for scpca ids not in 10x or vice versa?" But, I think you've written everything clearly! Just my brain keeping track, albeit late in a long day.... 

Either way, maybe a slightly more informative header here (and/or subheaders below?) might help future brains stay on top of what's going on.

sjspielman commented on data-raw/build_gene_reference.Rmd:
````diff
@@ -0,0 +1,179 @@
+  )
+```
+
+More than I expected, but it looks like most of these are cases where the gene symbol has been updated.
````
I see about 900 rows here... how did you make this conclusion about "most cases"?

sjspielman commented on data-raw/build_gene_reference.Rmd:
````diff
@@ -0,0 +1,179 @@
+```
+
+More than I expected, but it looks like most of these are cases where the gene symbol has been updated.
+What we are really interested in is the cases where the process of making the gene symbol unique had different effects:
````
```suggestion
What we are really interested in are the cases where the process of making the gene symbol unique had different effects:
```

sjspielman replied:
Can you elaborate just a bit on "different effects"?

sjspielman reviewed (commented):
I've done a _super cursory_ review here with some initial thoughts, but I've done so at the end of the day and I very much plan to have look with fresh eyes tomorrow! Some of my reviews can also be described as "how does someone without fresh eyes approach the notebook?"


And for now, one question:

> I also store the table only for internal use; I suppose I could make it available outside the package as well, but I wasn't sure whether that would be worth the effort. Maybe it would?

Wondering where else you might envision.. like a public S3 bucket?

jashapiro commented:
> And for now, one question:
> 
> > I also store the table only for internal use; I suppose I could make it available outside the package as well, but I wasn't sure whether that would be worth the effort. Maybe it would?
> 
> Wondering where else you might envision.. like a public S3 bucket?

What I meant here was that the way I include the table `scpca_gene_reference` in the package means that it is an internal object. It could be accessed as `rOpenScPCA:::scpca_gene_reference` but not with only `::` or by loading the package with `library()`.

sjspielman commented:
> What I meant here was that the way I include the table scpca_gene_reference in the package means that it is an internal object. It could be accessed as rOpenScPCA:::scpca_gene_reference but not with only :: or by loading the package with library().

Aha. It might be worth exporting since it is a useful piece of information but that said I can't immediately think of applications besides gene conversion as performed within rOpenScPCA here. Maybe something to circle back to if/when it becomes clear this information is otherwise helpful?

sjspielman commented on R/convert-gene-ids.R:
```diff
@@ -71,6 +92,10 @@ ensembl_to_symbol <- function(ensembl_ids, sce, leave_na = FALSE) {
+#' `10x2020`, or `10x2024`. If `sce` (the default) the internal row data is used.
+#' @param unique Whether to use unique gene symbols, as would be done if
+#' data had been read in with gene symbols by Seurat. Default is FALSE.
 #' @param convert_hvg Logical indicating whether to convert highly variable genes to gene symbols.
 #' @param convert_pca Logical indicating whether to convert PCA rotation matrix to gene symbols.
```
Indicate here that default is `TRUE`.

sjspielman commented on data-raw/build_gene_reference.nb.html:
I don't think this was fully regenerated.. Screenshot below.

Personally I find that rendering with `Preview` in RStudio tends to sometimes not work when I expect it to due to caching and/or order of running chunks, so I try (try is doing a lot of work here!) to make sure to render notebooks with `rmarkdown::render()`. This may or may not be relevant to you, so worst case this is another reminder to me to try to make sure to do that :)

As I write this I also wonder if actually we do want to split this up into two parts: a script to download and add the internal data, the notebook for exploration, and then add a new shell script to run them both? 


------------------


![Screenshot 2024-11-27 at 11 42 22 AM](https://github.com/user-attachments/assets/85a86896-0390-43a7-8b5c-6d1be0e02309)

jashapiro commented:
@sjspielman I updated this PR to split the script to build the data from the notebook to evaluate it. I believe I also addressed all of your comments, adding a message when an SCE is being used as the reference, fixing spacing, and updating docs. Should be ready for another look.

sjspielman commented on R/data.R:
👍

sjspielman commented on data-raw/README.md:
```diff
@@ -0,0 +1,11 @@
+## Building Gene References
+
+- The `build_gene_references.R` script creates a table of Ensembl id to gene symbol references.
+It starts with the example ScPCA-formatted SCE object stored in test data, extracting the table of gene ids and gene symbols.
+This is combined with the reference information extracted from example 10x Genomics datasets.
+The full table is saved in `data/scpca_gene_reference.rda`.
```
```suggestion
The script extracts the table of gene ids and gene symbols from the example ScPCA-formatted SCE object and combines it with the reference information extracted from example 10x Genomics datasets.
The full table is saved in `data/scpca_gene_reference.rda`.
```

sjspielman replied:
take it or leave it, your text is also fine.

actually though, should we say "example sce" or "test sce"?

sjspielman replied:
Might also say that re-running this script will overwrite the data

jashapiro replied:
Updated the text

sjspielman commented on data-raw/explore_gene_reference.Rmd:
````diff
@@ -0,0 +1,120 @@
+```
+
+
+# Load the R data
````
```suggestion
# Load the R data

```

sjspielman commented on data-raw/explore_gene_reference.Rmd:
```diff
@@ -0,0 +1,120 @@
+
+### Disagreements between symbols
+
+One other question is how often the `unique` gene symbols disagree for the same Ensemble ID.
```
```suggestion
One other question is how often the `unique` gene symbols disagree for the same Ensembl ID.
```

sjspielman commented on data-raw/explore_gene_reference.Rmd:
````diff
@@ -0,0 +1,120 @@
+  count(same_symbol)
+```
+
+What we are really interested in is the cases where the process of making the gene symbol unique had different effects:
````
```suggestion
What we are really interested in are the cases where the process of making the gene symbol unique had different effects:
```

sjspielman commented on data-raw/explore_gene_reference.Rmd:
````diff
@@ -0,0 +1,120 @@
+  )
+```
+
+This definitely happens more than I would have hoped, though my hope was probably unreasonable, as 14 cases is not bad.
````
14 seems pretty good tbh, or maybe my hopes were unreasonable in the _opposite_ direction 😄

sjspielman commented on data-raw/explore_gene_reference.Rmd:
I'd add some text at the top of this notebook to explicitly state as a reminder that you're looking at 2020, not 2024, throughout, because 2020 wasn't using Ensembl yet

sjspielman replied:
Or, is it also worth adding a section to confirm that 2024 and scpca are already as similar as we expect, since we have a notebook anyways?

jashapiro replied:
I didn't do that at this stage because I couldn't actually find any real datasets that use the new reference. I don't think this level of analysis is necessary at this time.

sjspielman commented on tests/testthat/test-convert-gene-ids.R:
```diff
@@ -1,30 +1,86 @@
-# read in SCE for testing
-sce <- readRDS(test_path("data", "scpca_sce.rds"))
```
I wonder whether it's starting to get worthwhile to have a [`setup.R`](https://testthat.r-lib.org/articles/special-files.html#setup-files) script that at least load the sce file? Separate PR, if at all, though.

sjspielman reviewed (commented):
I left a few more small comments, but this is about there! No big comments at this point.

jashapiro reviewed (commented):
@sjspielman I believe I addressed your comments.

sjspielman commented on data-raw/explore_gene_reference.Rmd:
```diff
@@ -0,0 +1,120 @@
+---
+title: "Gene ID Conversion data setup"
```
exploration, not setup

sjspielman reviewed (approved):
👍 
Just two things (but I don't need to see again!):
- I do feel fairly strongly that the notebook and or the `data-raw/README.md` need(s) to plainly state that the notebook is only comparing scpca and the 2020 10x reference, not 2024. Currently, the first notebook text makes it sound like you're setting up to do more comparisons ("but some of the 10x references have Ensembl IDs not present in the ScPCA data."), and similarly the README says "among" (vs between) when only 2 are being compared. 
- The notebook needs a new title since it is no longer setting up the conversion which was moved to the script.
